lib/ClangImporter/ClangIncludePaths.cpp

```diff
@@ -206,8 +206,18 @@ getLibcFileMapping(ASTContext &ctx, StringRef modulemapFileName,
   // Ideally we would check that all of the headers referenced from the
   // modulemap are present.
   Path libcDir;
-  if (auto dir = findFirstIncludeDir(
-          parsedIncludeArgs, {"inttypes.h", "unistd.h", "stdint.h"}, vfs)) {
+  if (triple.isAndroid() &&
+      !clangDriverArgs.getLastArgValue(clang::driver::options::OPT__sysroot_EQ)
+           .empty()) {
+    // Swift's driver passes in the Android NDK path using the --sysroot Clang
+    // flag.
+    libcDir =
+        clangDriverArgs.getLastArgValue(clang::driver::options::OPT__sysroot_EQ)
```

Comment:
Was this really necessary? I made sure [the `sdkPath` is passed in to the clang Driver above](https://github.com/apple/swift/blob/01acb5680f1396b20c5805c0173f07f19c2ed972/lib/ClangImporter/ClangIncludePaths.cpp#L177), been working fine for me with `findFirstIncludeDir()` below.

Reply:
The `-sdk` option will be used for the SDK overlay much like we do so on Windows.

Reply:
Yes, we are doing something different. `-sdk` provides the path to the "Swift SDK" (e.g. `%LocalAppData%\Programs\Swift\Platforms\Android.platform\Developer\SDKs\Android.sdk`). The `--sysroot` will be the path to the NDK that is overlaid with the "Swift SDK". The user is responsible for providing the NDK, but we will distribute the SDK for Android with the toolchain.

Reply:
I think you misunderstand what the Swift compiler considers a SDK on these Unix platforms (I can't speak for Darwin or Windows, as I've never built Swift on those platforms). The SDK _is_ the C sysroot for that OS, taken to be `/` by default. If the compiler finds the Swift resource directory in the SDK at `/usr/lib/swift/<os>/`, it uses that; otherwise, you must specify one with `-resource-dir`.

That is how the Android NDK config I linked above works: there is no need to add other paths like this or change what paths you are passing in. The `clangDriver.SysRoot` I linked above already sets the C sysroot based on the `-sdk` passed in.

Why would you want to pass in your `Android.sdk/` directory instead to `-sdk`? I don't know what that contains. Is your Swift resource directory at `Android.sdk/usr/lib/swift/android/`? If so, passing in `-sdk ndk/blah/blah/sysroot/ -resource-dir Android.sdk/usr/lib/swift/android/` should work fine, without needing special Android logic like this.

Reply:
The compiler resources are part of the compiler, not the SDK. The SDK contains the Swift modules and auxiliary data for the compiler that is target dependent. It also contains development content such as XCTest. The `--sysroot` bits are for clang and that will be from the NDK, which is not re-distributed by the toolchain.

Reply:
Right, the Swift compiler expects the C headers and libraries for the target OS to be in the SDK at `Android.sdk/usr/include/` and `Android.sdk/usr/lib/`, while it looks for the "the Swift modules and auxiliary data for the compiler that is target dependent" to be in `Android.sdk/usr/lib/swift/`.

Since the Android NDK only has a C/C++ sysroot, without any Swift files, the way Swift has always handled this is to pass the target OS's C sysroot as the `-sdk` and the path to the Swift files as the Swift `-resource-dir` (note the usr/lib/swift in the help like I said above):
```
> swiftc -help-hidden|ag resource
  -resource-dir </usr/lib/swift>
                          The directory that holds the compiler resource files
```
Now, you may be right about the terminology being incorrect, but this is long-standing practice that it would be better not to hack like this. If you want to change it, talk to the Swift team about updating terminology and flags for all platforms, which I'd be fine with.

@etcwilde or @MaxDesiatov, any input here?

Reply:
I have many conflicted feelings and no strong opinions beyond "we're all wrong", so here's my braindump. Yes, the resource-dir should be shipped as a compiler-locked part of the toolchain (like compiler-rt and the clang builtin headers for clang). Swift is terribly unclear about what is and isn't locked to that exact compiler. SwiftShims, that should be compiler-locked since there is some influence on the ABI generated from the compiler (mostly sizes and such). swiftrt.o is certainly locked to the compiler (if we ever wanted to add a section for a new kind of runtime metadata, that would need to go in there). Is the stdlib locked to a specific compiler build though? Swiftmodules are, which makes things interesting because that means everything is locked to a specific compiler. I don't love that idea though....

The fact that we have four flags that are needed in order to set up a Windows build to find all of the SDKs smells like we don't have the right abstraction somewhere (or does clang have something for `-windows-sdk-root`, `-windows-sdk-version`, `-visualc-tools-root`, and `-visualc-tools-version`?). My initial thought it that maybe we could have `-sdk` behave a little like `-I` and `-L`, where repeating them allows you to search SDKs in a specific order looking for headers and libraries in each one until you find it (the library and headers for that library would need to come from the same SDK or a very sad day could occur, but it's an idea). The challenge here is that not all SDKs are laid out the same and I'm not sure if there's quite enough information in the triple to pull the right file layout info.

Looking at something like Rust, Rust runtimes aren't installed in SDKs or on platforms because the language has no ABI stability, either in runtime struct layout or in compiler calling conventions, so you really can't. At that point, everything is revlocked to the exact version of the compiler that built it, so to some extent, they're in the same spot as Swift with everything mixed. Rust points their sysroot at the location where the Rust binaries live, not where the system SDK lives. Unlike Rust though, Swift is able to expose a native developer experience for a given platform instead of stomping over it with a facsimile abstraction. Because of this, I don't believe they re-expose declarations from the libc that the Rust stdlib was built against since it's an implementation detail and not API. I kind of wonder if @artemcm's work on the explicit modules will help here? Implementation details of the Swift stdlib shouldn't leak up to clients. At that point, the SDK would just need to point at the location where the Swift standard library and swiftmodules live for that platform. Maybe the right thing is to `-I` and `-L` your way through until then? I don't know? ¯\\\_(ツ)_/¯

Reply:
A nice digression but not relevant to this question. This is a simple question: the Swift compiler expects both a C/C++ sysroot for a platform and Swift stdlib/corelibs for that same platform to be passed in a single `-sdk /path/to/Android.sdk/` path, then automatically looking in `/path/to/Android.sdk/usr/include/` for C/C++ headers, `/path/to/Android.sdk/usr/lib/` for C/C++ libraries, and in `/path/to/Android.sdk/usr/lib/swift/android/` for the Swift stdlib/corelibs. The question is how to handle the scenario where you have a separate C/C++ sysroot and Swift library path.

The way Swift has always handled this, including the toolchain build for linux and most every OS platform, is to pass `-sdk /path/to/c/sysroot/ -resource-dir /path/to/swift/usr/lib/swift/` instead. @hyp is instead hacking the compiler here to pass something like `-sdk /path/to/swift/ -Xcc --sysroot /path/to/c/sysroot/` instead, which seemingly requires the above Android hack. I am telling him and @compnerd that this is not the way it's done, and if they want to change it, they shouldn't hack Android alone but suggest a uniform change for all platforms.

Since both you and @MaxDesiatov have experience with Swift cross-compilation for these non-Darwin platforms, I'm asking if you want to chime in on this specific issue of how to handle split SDKs.

Reply:
Its exactly relevant to this question;

> My initial thought it that maybe we could have -sdk behave a little like -I and -L, where repeating them allows you to search SDKs in a specific order looking for headers and libraries in each one until you find it (the library and headers for that library would need to come from the same SDK or a very sad day could occur, but it's an idea). The challenge here is that not all SDKs are laid out the same and I'm not sure if there's quite enough information in the triple to pull the right file layout info.

Reply:
> Its exactly relevant to this question

As you yourself noted, that would not work well because you cannot know which header might come out of which SDK, whereas specifying the `-resource-dir` at least tells the compiler exactly where the Swift files are.

> I would expect the -sdk to point at the Swift stdlib once things are said and done and the -resource-directory to point at the Swift compiler resources.

That's fine if you two would like to make that change, but that's not the way it works on most any platform now, ie the `-sdk` points at the C sysroot and the `-resource-dir` points at the Swift stdlib.

How do you propose to change that?

Reply:
> That's fine if you two would like to make that change, but that's not the way it works on most any platform now, ie the -sdk points at the C sysroot and the -resource-dir points at the Swift stdlib.

I'm aware that this is the case on Linux. 

>> I have many conflicted feelings and no strong opinions beyond "we're all wrong".

"Historic reasons" is one of my conflicted feelings on this topic.
Windows and Darwin don't do this though. The SDK points at the Swift bits, but then that's why we need to point Swift at the Windows SDK. I think we need to generalize that. Darwin makes it hard to distinguish because Swift and the C/C++ bits live together.

From a user's perspective, why would I need to point Swift at the C/C++ SDK? It's Swift, it should just need the Swift interfaces/modules and libraries. If the Swift stdlib didn't expose the fact that it needed to find the C/C++ headers in order to compile against it, we wouldn't need it. Then if you _do_ want to pull in a C/C++ library, it would be up to the user to pass the appropriate `-I` and `-L` to find the headers and libraries for C/C++.

One idea might be `-c-sysroot`, or `-system-sdk` in addition to `-sdk` for the Swift parts? Then we key off of parts of the triple to determine what the expected system SDK layout is? If the system SDK isn't set, then it just defaults to whatever `-sdk` is. This might actually have a chance of being backward compatible.

Reply:
> I'm aware that this is the case on Linux... "Historic reasons" is one of my conflicted feelings on this topic.
Windows and Darwin don't do this though. The SDK points at the Swift bits, but then that's why we need to point Swift at the Windows SDK.

Alright, perhaps I overstated where it is done that way, but I noted I only have experience with Unix platforms.

> One idea might be -c-sysroot, or -system-sdk in addition to -sdk for the Swift parts? Then we key off of parts of the triple to determine what the expected system SDK layout is? If the system SDK isn't set, then it just defaults to whatever -sdk is. This might actually have a chance of being backward compatible.

Seems like a good idea to me, and if one of you proposes a pull implementing it, I would be happy to review it. I'm just against the Android-specific hack proposed here.

Reply:
I agree. This is a change of behavior and I would like to see an evolution proposal before we go through with it. I'm happy to help co-author on it if someone wants to put together a gist or create a branch in their fork.

Ultimately, I think we need to have a system agnostic way to say "this is where the system libraries and headers are" (new thing) and separate that from "this is where the Swift libraries and modules are" (`-sdk`) which is also separate from "these are the files that are specific to this Swift compiler" (`-resource-dir`). And perhaps the help-text should be clearer and actually helpful for what these flags mean. "The directory that holds the compiler resource files" is not particularly helpful.

Reply:
I like the `-c-sysroot` / `-c-sdk` flag idea!

Reply:
Adding to all of this, we should make sure that it's _possible_ to embed the C sysroot directly into the Swift SDK, such that you'd only need the `-sdk` option. We likely won't be doing that in many cases because of licensing issues, but we shouldn't rule it out.

Reply:
It's the other way around: the compiler looks in the C sysroot specified by `-sdk` at `<c-sysroot>/usr/lib/swift/` for the Swift libraries and module files, so if you have edit access to your C sysroot, you can always just symlink the two, ie `ln -s /path/to/swift-sdk/usr/lib/swift <c-sysroot>/usr/lib/swift`. This split SDK is only ever an issue if you cannot easily modify your C sysroot, say if it is in `/` and you don't have sudo access.

Reply:
I think we're talking at crossed purposes. What I mean is that it might be useful sometimes to ship e.g. an `Android.sdk` bundle that contains _all_ the files, as opposed to having to point `-sdk` at `Android.sdk` and then `-c-sysroot` or whatever we're calling it at the downloaded NDK.

(I'm not suggesting we _do_ that for Android. Just that it might be a good idea.)

Reply:
Unfortunately, google has a weird license on the NDK that includes this clause, "Except to the extent required by applicable third party licenses, you may not copy (except for backup purposes), modify, adapt, redistribute, decompile, reverse engineer, disassemble, or create derivative works of the SDK or any part of the SDK." Click on [one of the file downloads here to see it](https://developer.android.com/ndk/downloads).

It is probably legally unenforceable since AFAIK their NDK is pure OSS, they have not enforced that distribution clause AFAIK, it is probably legal boilerplate they simply copied over from their main Java SDK, but it would probably stop most from trying.

My symbolic link workaround gets around that.

Reply:
I think he was talking hypothetically. Like if Google ever decided to ship Swift in their NDK, or if Ubuntu/Debian shipped the Swift runtimes as an apt package, then you made a sysroot out of that Ubuntu/Debian install, you should only need to pass `-sdk` and Swift will find both the C, C++, and Swift libraries at that location appropriately. That's where my thought was to default the `-system-sdk` to `-sdk` if it were left unspecified.

Reply:
Oh, that's already the default, as I stated multiple times in this comment thread. However, I just checked and while this C++ Driver does that, the new swift-driver does not carry that over. I will file an issue for that discrepancy now.

Reply:
Filed, apple/swift-driver#1562.